test/extended/router/h2spec.go

```diff
@@ -55,47 +53,89 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+
+			g.By("Locating the router image reference")
 			routerImage, err := exutil.FindRouterImage(oc)
 			o.Expect(err).NotTo(o.HaveOccurred())
 
-			err = oc.Run("new-app").Args("-f", configPath, "-p", "HAPROXY_IMAGE="+routerImage).Execute()
+			g.By("Locating the canary image reference")
+			canaryImage, err := getCanaryImage(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating h2spec test service")
+			err = oc.Run("new-app").Args("-f", h2specServiceConfigPath,
+				"-p", "HAPROXY_IMAGE="+routerImage,
+				"-p", "H2SPEC_IMAGE="+canaryImage).Execute()
 			o.Expect(err).NotTo(o.HaveOccurred())
 
 			e2e.ExpectNoError(e2epod.WaitForPodNameRunningInNamespace(oc.KubeClient(), "h2spec-haproxy", oc.KubeFramework().Namespace.Name))
+			e2e.ExpectNoError(e2epod.WaitForPodNameRunningInNamespace(oc.KubeClient(), "h2spec", oc.KubeFramework().Namespace.Name))
+
+			shardFQDN := oc.Namespace() + "." + defaultDomain
 
-			routeTypeTests := []h2specRouteTypeTest{{
-				routeType: routev1.TLSTerminationEdge,
-				knownFailures: map[string]bool{
-					"http2/4.2.3": true,
-				},
-			}, {
-				routeType: routev1.TLSTerminationPassthrough,
-				knownFailures: map[string]bool{
-					"http2/4.2.3": true,
-				},
-			}, {
-				routeType: routev1.TLSTerminationReencrypt,
-				knownFailures: map[string]bool{
-					"http2/4.2.3": true,
-				},
-			}}
-
-			g.By("verifying accessing the host returns a 200 status code")
-			for i := 0; i < len(routeTypeTests); i++ {
-				urlTester := url.NewTester(oc.AdminKubeClient(), oc.KubeFramework().Namespace.Name).WithErrorPassthrough(true)
-				defer urlTester.Close()
-				hostname, err := getHostnameForRoute(oc, fmt.Sprintf("h2spec-haproxy-%s", routeTypeTests[i].routeType))
+			// The new router shard is using a namespace
+			// selector so label this test namespace to
+			// match.
+			g.By("By labelling the namespace")
+			err = oc.AsAdmin().Run("label").Args("namespace", oc.Namespace(), "type="+oc.Namespace()).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating routes to test for h2spec compliance")
+			err = oc.Run("new-app").Args("-f", h2specRoutesConfigPath,
+				"-p", "DOMAIN="+shardFQDN,
+				"-p", "TYPE="+oc.Namespace()).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating a test-specific router shard")
+			shardConfigPath, err = shard.DeployNewRouterShard(oc, 10*time.Minute, shard.Config{
+				FixturePath: h2specRouterShardConfigPath,
+				Domain:      shardFQDN,
+				Type:        oc.Namespace(),
+			})
+			o.Expect(err).NotTo(o.HaveOccurred(), "new router shard did not rollout")
+
+			g.By("Getting LB service")
+			shardService, err := getRouterService(oc, 10*time.Minute, "router-"+oc.Namespace())
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(shardService).NotTo(o.BeNil())
+			o.Expect(shardService.Status.LoadBalancer.Ingress).ShouldNot(o.BeEmpty())
+
+			var addrs []string
+
+			if len(shardService.Status.LoadBalancer.Ingress[0].Hostname) > 0 {
```

Comment:
Doesn't have to be now but at some point we should consider creating a helper function for this "Ingress Controller DNS resolves as expected" logic.

Reply:
Yep, agreed. And I started that and backed out. :)